src/NHibernate/Linq/Visitors/NhPartialEvaluatingExpressionVisitor.cs

```diff
@@ -155,6 +158,42 @@ private Expression EvaluateSubtree(Expression subtree)
 
 		#region NH additions
 
+		protected override Expression VisitMethodCall(MethodCallExpression node)
+		{
+			DetectEvaluatableExpressionOnCollectionContains(node);
+			return base.VisitMethodCall(node);
+		}
+
+		private void DetectEvaluatableExpressionOnCollectionContains(MethodCallExpression expression)
+		{
+			if (!expression.Method.IsGenericMethod || ContainsMethodInfo != expression.Method.GetGenericMethodDefinition())
+				return;
+			var argument = expression.Arguments[0];
+			if (argument.NodeType != ExpressionType.Call)
+				return;
+
+			if(TryGetCollectionParameter((MethodCallExpression)argument))
+				_partialEvaluationInfo.AddEvaluatableExpression(argument);
+		}
+
+		private bool TryGetCollectionParameter(MethodCallExpression expression)
+		{
+			if (expression.Method.DeclaringType != typeof(Enumerable))
+				return IsCollectionParameter(expression);
+
+			var arg = expression.Arguments[0];
+
+			if (IsCollectionParameter(arg))
+				return true;
+
+			return arg.NodeType == ExpressionType.Call && TryGetCollectionParameter((MethodCallExpression) arg);
```

Comment:
Why do we need to do that recursively?

Reply:
To support chain of LINQ commands. And to make sure it doesn't start with mapped association (somethin like
`db.Orderes.Where( x => x.OrderLines.Select(...).Where(...).Contains(...))` should not be evaluated)